### PR TITLE
Updated SQuAD pipeline

### PR DESCRIPTION
**Pull request by natuan**
Updated the SQuAD pipeline to be closer to the eval from SparseML training pipeline.

Added two options for profiling and saving predictions (n-best, final) that could be used for debugging.

### CODE REVIEW

**Review by anmarques — DISMISSED** [review]

File: src/deepsparse/transformers/pipelines/question_answering.py [hunk 1]
```diff
@@ -63,6 +63,7 @@ class QuestionAnsweringInput(BaseModel):
 
     question: str = Field(description="String question to be answered")
     context: str = Field(description="String representing context for answer")
+    id: str = None
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/question_answering.py:66]: can we make the id a `Field` with `default=None` instead?

File: src/deepsparse/transformers/pipelines/question_answering.py [hunk 2]
```diff
@@ -209,18 +255,35 @@ def process_inputs(
...
+            span_input = [
+                numpy.array(tokenized_example[k][i]) for k in self.onnx_input_names
+            ]
+            span_engine_inputs.append(span_input)
+
+            span_extra_info.append(
+                {
+                    k: numpy.array(tokenized_example[k][i])
+                    for k in tokenized_example.keys()
+                    if k not in self.onnx_input_names
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/question_answering.py:275]: nit: prefer to rename `i -> span` `k -> key`

**Review by spacemanidol — DISMISSED** [review]

File: src/deepsparse/transformers/pipelines/question_answering.py [hunk 3]
```diff
@@ -233,80 +296,219 @@ def process_engine_outputs(
...
+                    float(v)
+                    if isinstance(v, (numpy.float16, numpy.float32, numpy.float64))
+                    else v
+                )
+                for k, v in pred.items()
+            }
+            for pred in predictions
+        ]
+
+        if self.output_dir is not None:
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/question_answering.py:481]: let's move this section to a helper function to maintain the flow in postprocessing

File: src/deepsparse/transformers/pipelines/question_answering.py [hunk 4]
```diff
@@ -324,145 +526,56 @@ def route_input_to_bucket(
...
-                        tok != 1 if question_first else 0
-                        for tok in encoded_inputs.sequence_ids(0)
-                    ]
+            # For evaluation, we will need to convert our predictions to substrings of
+            # the context, so we keep the corresponding example_id and we will store
+            # the offset mappings.
+            tokenized_example["example_id"] = []
+
+            n_spans = len(tokenized_example["input_ids"])
+            for i in range(n_spans):
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/question_answering.py:559]: nit: i -> `span`

**Review by bfineran — COMMENTED** [review]
great work @natuan, few quick comments

**Comment by bfineran** [comment]
additionally, do we have any number on the performance of the updated pipeline vs the previous?

File: src/deepsparse/transformers/pipelines/pipeline.py [hunk 5]
```diff
@@ -135,7 +135,7 @@ def tokens_to_engine_input(
         if not all(name in tokens for name in self.onnx_input_names):
             raise ValueError(
                 f"pipeline expected arrays with names {self.onnx_input_names}, "
-                f"received inputs: {list(tokens.keys())}"
+                f"received inputs: {list(tokens[0].keys())}"
```
[the remark this answers is not in the document]
Comment (natuan) [reply src/deepsparse/transformers/pipelines/pipeline.py:138]: I'm not sure, will update if it fails with other pipelines.

**Comment by natuan** [comment]
> additionally, do we have any number on the performance of the updated pipeline vs the previous?

Posted this morning: https://neuralmagic.slack.com/archives/C01F0PWAPFU/p1657512157198479

File: src/deepsparse/transformers/pipelines/pipeline.py (hunk 5 above)
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/pipeline.py:138]: bumping this - the if check expects tokens to be a dict here, but we seem to be expecting a list in the error message. should we actually be unwrapping a potential list before the if statement?

Comment (natuan) [reply src/deepsparse/transformers/pipelines/pipeline.py]: Actually my new code no longer calls into this func, so just reverted it

**Review by bfineran — APPROVED** [review]

**Review by anmarques — APPROVED** [review]

**Pull request merged**